### PR TITLE
Added RethinkDB external pillar module

### PR DESCRIPTION
**Pull request by michelsen**
### What does this PR do?
This pull request add an external pillar module for RethinkDB

### What issues does this PR fix or reference?
Adds external pillar module for RethinkDB. Does not reference anything.

### Tests written?
No

### CODE REVIEW

File: salt/pillar/rethinkdb_pillar.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    if data.items:
+
+        # Return nothing if multiple documents are found for a minion
+        if len(data.items) > 1:
+            log.error('ext_pillar.rethinkdb: ambiguous documents found for '
+                      'minion {0}'.format(minion_id))
+            return {}
+
+        else:
+            for document in data:
```
Comment (cachedout) [inline salt/pillar/rethinkdb_pillar.py]: I'm not sure I totally follow the reasoning for this loop. The way I read this, if it's > 1 then the above block will execute but if it's 1 or 0, then we'll go here but it's not clear why we're looping for 1 | 0 items. Am I missing something here?

**Review by cachedout — COMMENTED** [review]
There's just one bit here that I'm not quite following. If you could help clarify what I'm not seeing here then I'm sure we can get this merged. Thanks!

File: salt/pillar/rethinkdb_pillar.py (hunk 1 above)
Comment (michelsen) [reply salt/pillar/rethinkdb_pillar.py]: You're quite correct. I was going by the RethinkDB examples I read some time ago that dealt with iterating over results in a cursor. I updated the pull request and replaced the for loop with pop()

File: salt/pillar/rethinkdb_pillar.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
+# -*- coding: utf-8 -*-
+'''
+Provide external pillar data from RethinkDB
+
```
Comment (rallytime) [inline salt/pillar/rethinkdb_pillar.py]: Can you add `.. versionadded:: Oxygen` here?

File: salt/pillar/rethinkdb_pillar.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+In the example above the following happens.
+    * The salt-master will look for external pillars in the 'ext_pillar' table
+      on the RethinkDB host
+    * The minion id will be matched against the 'minion_id' field
+    * Pillars will be retrieved from the nested field 'pillar_root'
+    * Found pillars will be merged inside a key called 'external_pillar'
+
+
+Module Documentation
+====================
```
Comment (rallytime) [inline salt/pillar/rethinkdb_pillar.py]: Did you intend to add some documentation here? If not, then I think this section header should be removed.

File: salt/pillar/rethinkdb_pillar.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+def ext_pillar(minion_id,
+               pillar,
+               table='pillar',
+               id_field=None,
+               field=None,
+               pillar_key=None):
+    '''
+    Collect minion external pillars from a RethinkDB database
+
+Arguments:
```
Comment (rallytime) [inline salt/pillar/rethinkdb_pillar.py]: This line should be indented to match up with the rest of the doc section.

File: salt/pillar/rethinkdb_pillar.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    port = __opts__['rethinkdb.port']
+    database = __opts__['rethinkdb.database']
+    username = __opts__['rethinkdb.username']
+    password = __opts__['rethinkdb.password']
+
+    log.debug('Connecting to {0}:{1} as user \'{2}\' for RethinkDB ext_pillar'
+              .format(host, port, username))
+
+    # Connect to the database
+    conn = r.connect(host=host,
```
Comment (rallytime) [inline salt/pillar/rethinkdb_pillar.py]: I know this is picky, but can you change this to use a three-letter variable? That's what we like to use according to Salt's [style guide](https://docs.saltstack.com/en/latest/topics/development/conventions/style.html#variables).

**Review by rallytime — CHANGES_REQUESTED** [review]
Hi @michelsen - I just had a couple of small things that need to be cleaned up. Then I think we can get this in.

**Comment by rallytime** [comment]
@michelsen Welcome to Salt! This looks great overall. I had a couple of small salt-specific requests. Thank you very much for your contribution.

**Comment by cachedout** [comment]
@michelsen Did you see the above requests for some minor changes? We can't merge this until those are in so please let me know when you're able to follow up with those. Thanks!

**Comment by michelsen** [comment]
I

**Comment by michelsen** [comment]
Ugh, commenting on my phone and accidentally closed the pull request ...
I have read and agree with the comments so far, I am however indisposed this week and will incorporate the proposed changes when I am back this Sunday.

File: salt/pillar/rethinkdb_pillar.py (hunk 3 above)
Comment (michelsen) [reply salt/pillar/rethinkdb_pillar.py]: I took some inspiration from the mongodb external pillar module, where I liked how the function docstrings have a section header above them which makes it look more organized on the docs page. If it is bad practice, then I don't mind removing it.

File: salt/pillar/rethinkdb_pillar.py (hunk 5 above)
Comment (michelsen) [reply salt/pillar/rethinkdb_pillar.py]: Absolutely, this is just a habit picked up from the rethinkdb code examples. I should have caught this. I will fix this when I'm back at my keyboard this Sunday.

File: salt/pillar/rethinkdb_pillar.py (hunk 3 above)
Comment (rallytime) [reply salt/pillar/rethinkdb_pillar.py]: Ah, ok. I think that is fine. I was just confused why there weren't docs there, but that makes sense when the docs are built out.

File: salt/pillar/rethinkdb_pillar.py (hunk 5 above)
Comment (rallytime) [reply salt/pillar/rethinkdb_pillar.py]: Sounds good - Thank you!

**Comment by cachedout** [comment]
@michelsen Thank you very much. We look forward to getting this merged!

File: salt/pillar/rethinkdb_pillar.py (hunk 2 above)
Comment (michelsen) [reply salt/pillar/rethinkdb_pillar.py]: Done

File: salt/pillar/rethinkdb_pillar.py (hunk 4 above)
Comment (michelsen) [reply salt/pillar/rethinkdb_pillar.py]: Done

File: salt/pillar/rethinkdb_pillar.py (hunk 5 above)
Comment (michelsen) [reply salt/pillar/rethinkdb_pillar.py]: I removed the shorthand import name since it's not really necessary.

**Review by michelsen — COMMENTED** [review]
Done

**Review by rallytime — APPROVED** [review]

**Review by cachedout — APPROVED** [review]

**Comment by garethgreenaway** [comment]
re-run ubuntu

**Pull request merged**